core/server/api_container/server/startosis_engine/plan_yaml/plan_yaml_generator.go

```diff
@@ -22,38 +22,56 @@ const (
 	outputFutureRefType    = "output"
 )
 
-// PlanYaml is a yaml representation of the effect of an Instructions Plan or sequence of instructions on the state of the Enclave.
-type PlanYaml struct {
-	privatePlanYaml *privatePlanYaml
+// PlanYamlGenerator is a yaml representation of the effect of an Instructions Plan or sequence of instructions on the state of the Enclave.
+type PlanYamlGenerator struct {
+	privatePlanYaml *PlanYaml
 
 	futureReferenceIndex map[string]string
 	filesArtifactIndex   map[string]*FilesArtifact
 	latestUuid           int
+	imageSet             map[string]bool
 }
 
-func CreateEmptyPlan(packageId string) *PlanYaml {
-	return &PlanYaml{
-		privatePlanYaml: &privatePlanYaml{
-			PackageId:      packageId,
-			Services:       []*Service{},
-			Tasks:          []*Task{},
-			FilesArtifacts: []*FilesArtifact{},
+func CreateEmptyPlan(packageId string) *PlanYamlGenerator {
+	return &PlanYamlGenerator{
+		privatePlanYaml: &PlanYaml{
+			PackageId:           packageId,
+			Services:            []*Service{},
+			Tasks:               []*Task{},
+			FilesArtifacts:      []*FilesArtifact{},
+			Images:              []string{},
+			PackageDependencies: []string{},
 		},
+		imageSet:             map[string]bool{},
 		futureReferenceIndex: map[string]string{},
 		filesArtifactIndex:   map[string]*FilesArtifact{},
 		latestUuid:           0,
 	}
 }
 
-func (planYaml *PlanYaml) GenerateYaml() (string, error) {
-	yamlBytes, err := yaml.Marshal(planYaml.privatePlanYaml)
+func (planYaml *PlanYamlGenerator) GenerateYaml(dependenciesOnly bool) (string, error) {
+	planYaml.AddImages()
+
+	var planYamlCopy PlanYaml
+	planYamlCopy = *planYaml.privatePlanYaml
+	if dependenciesOnly {
+		planYamlCopy = PlanYaml{
+			PackageId:           "",
```

Comment:
I was confused for a few seconds by why we're nulling out the package ID, as well as a bunch of other stuff. I'm guessing it's because we expect this YAML to be consumed by the CLI, which will process the images & dependencies for the `--pull` flag?

I think my expectation would have been that the plan YAML we return to the CLI is always complete, and either the user or the CLI can consume whatever sub-parts it needs (rather than us suppressing info on the APIC end).

Rationale: it seems like we're then tying the APIC's workings too much to what sub-parts of the output the user might want (e.g. then we'd be opened up to the APIC needing to change if the user wants a `--services-only` flag). 

Would it be feasible to have the PlanYamlGenerator always produce the full YAML, and then the CLI or the user can just pull out the parts they want?

Reply:
agreed, i think i did the initially bc it was faster than processing the yaml on the CLI said but yes